bindings/java/src/main/com/apple/foundationdb/tuple/TupleUtil.java

```diff
@@ -159,318 +261,325 @@ static int getCodeFor(Object o) {
+			byte[] range;
+			if(state.nullCount == 0) {
+				range = Arrays.copyOfRange(rep, start, end);
+			}
+			else {
+				ByteBuffer dest = ByteBuffer.allocate(end - start - state.nullCount);
+				ByteArrayUtil.replace(rep, start, end - start, NULL_ESCAPED_ARR, NULL_ARR, dest);
+				range = dest.array();
+			}
 			//System.out.println(" -> byte string contents: '" + ArrayUtils.printable(range) + "'");
-			return new DecodeResult(end + 1, range);
+			state.add(range, end + 1);
 		}
-		if(code == STRING_CODE) {
-			int end = ByteArrayUtil.findTerminator(rep, (byte)0x0, (byte)0xff, start, last);
+		else if(code == STRING_CODE) {
+			int end = state.findNullTerminator(rep, start, last);
 			//System.out.println("End of UTF8 string: " + end);
-			byte[] stringBytes = ByteArrayUtil.replace(rep, start, end - start, NULL_ESCAPED_ARR, new byte[] { nil });
-			String str = new String(stringBytes, UTF8);
+			String str;
+			if(state.nullCount == 0) {
+				str = new String(rep, start, end - start, UTF8);
+			}
+			else {
+				ByteBuffer dest = ByteBuffer.allocate(end - start - state.nullCount);
+				ByteArrayUtil.replace(rep, start, end - start, NULL_ESCAPED_ARR, NULL_ARR, dest);
+				str = new String(dest.array(), UTF8);
+			}
 			//System.out.println(" -> UTF8 string contents: '" + str + "'");
-			return new DecodeResult(end + 1, str);
+			state.add(str, end + 1);
 		}
-		if(code == FLOAT_CODE) {
-			byte[] resBytes = Arrays.copyOfRange(rep, start, start+4);
-			floatingPointCoding(resBytes, 0, false);
-			float res = ByteBuffer.wrap(resBytes).order(ByteOrder.BIG_ENDIAN).getFloat();
-			return new DecodeResult(start + 4, res);
+		else if(code == FLOAT_CODE) {
+			int rawFloatBits = ByteBuffer.wrap(rep, start, Float.BYTES).getInt();
+			float res = decodeFloatBits(rawFloatBits);
+			state.add(res, start + Float.BYTES);
 		}
-		if(code == DOUBLE_CODE) {
-			byte[] resBytes = Arrays.copyOfRange(rep, start, start+8);
-			floatingPointCoding(resBytes, 0, false);
-			double res = ByteBuffer.wrap(resBytes).order(ByteOrder.BIG_ENDIAN).getDouble();
-			return new DecodeResult(start + 8, res);
+		else if(code == DOUBLE_CODE) {
+			long rawDoubleBits = ByteBuffer.wrap(rep, start, Double.BYTES).getLong();
+			double res = decodeDoubleBits(rawDoubleBits);
+			state.add(res, start + Double.BYTES);
 		}
-		if(code == FALSE_CODE) {
-			return new DecodeResult(start, false);
+		else if(code == FALSE_CODE) {
+			state.add(false, start);
 		}
-		if(code == TRUE_CODE) {
-			return new DecodeResult(start, true);
+		else if(code == TRUE_CODE) {
+			state.add(true, start);
 		}
-		if(code == UUID_CODE) {
-			ByteBuffer bb = ByteBuffer.wrap(rep, start, 16).order(ByteOrder.BIG_ENDIAN);
+		else if(code == UUID_CODE) {
+			ByteBuffer bb = ByteBuffer.wrap(rep, start, 2 * Long.BYTES).order(ByteOrder.BIG_ENDIAN);
 			long msb = bb.getLong();
 			long lsb = bb.getLong();
-			return new DecodeResult(start + 16, new UUID(msb, lsb));
+			state.add(new UUID(msb, lsb), start + 16);
```

Comment:
Not a big deal, but you used 2*Long.BYTES above and a constant 16 here. You could make those the same to be more clear.

Reply:
I made a `UUID_BYTES` constant that is the number of bytes in a `UUID`. I then tried to replace all instances to 16 or `2 * Long.BYTES` with the new constant.